crates/ty_python_semantic/src/types/generics.rs

```diff
@@ -481,15 +589,60 @@ impl<'db> Specialization<'db> {
         type_mapping: &TypeMapping<'a, 'db>,
         visitor: &ApplyTypeMappingVisitor<'db>,
     ) -> Self {
+        let applied_specialization_type = type_mapping.get_specialization_type(db);
```

Comment:
This feels not great and I tried to implement it in `apply_specialization` instead, but couldn't get that to work for a reason I don't fully understand.

Reply:
What I tried was this:

```
--- a/crates/ty_python_semantic/src/types.rs
+++ b/crates/ty_python_semantic/src/types.rs
@@ -5971,7 +5971,11 @@ impl<'db> Type<'db> {
         db: &'db dyn Db,
         specialization: Specialization<'db>,
     ) -> Type<'db> {
-        self.apply_type_mapping(db, &TypeMapping::Specialization(specialization))
+        let new_specialization = self.apply_type_mapping(db, &TypeMapping::Specialization(specialization));
+        match specialization.specialization_type(db) {
+            None => new_specialization,
+            Some(materialization_type) => new_specialization.materialize(db, materialization_type)
+        }
     }
 
     fn apply_type_mapping<'a>(
```

Which should apply recursively since `materialize` is recursive.

Reply:
#20121

Reply:
It has something to do with how bases are specialized. Given this test code:

```py
def _(y: Top[InvariantChild[Any]]) -> None:
    reveal_type(y.__mro__)
```

In the working branch, this reveals `tuple[<class 'Top[InvariantChild[Any]]'>, <class 'CovariantBase[object]'>, typing.Generic, <class 'object'>]`, which is expected: the covariant base class materializes the typevar to `object`.

In the non-working branch, this reveals `tuple[<class 'Top[InvariantChild[Any]]'>, <class 'CovariantBase[object]'>, typing.Generic, <class 'object'>, <class 'CovariantBase[Any]'>]` -- for some reason we have both `CovariantBase[object]` and `CovariantBase[Any]` still in the MRO.

Reply:
I'm out of time to look into this for now; I think we should go ahead and land this with a TODO comment. I'll add the comment and then merge.